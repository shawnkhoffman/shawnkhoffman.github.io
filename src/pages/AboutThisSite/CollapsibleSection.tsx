import React, { useState, useId } from 'react';

interface CollapsibleSectionProps {
  title: string;
  content: string;
  testId?: string;
}

const CollapsibleSection: React.FC<CollapsibleSectionProps> = ({ 
  title, 
  content,
  testId = 'collapsible-section'
}) => {
  const [isOpen, setIsOpen] = useState(false);
  const uniqueId = useId();
  const contentId = `collapsible-content-${uniqueId}`;

  const handleToggle = (event: React.SyntheticEvent) => {
    event.preventDefault();
    setIsOpen(prev => !prev);
  };

  const handleKeyDown = (event: React.KeyboardEvent) => {
    if (event.key === 'Enter' || event.key === ' ') {
      event.preventDefault();
      handleToggle(event);
    }
  };

  const handleTouchEnd = (event: React.TouchEvent) => {
    event.preventDefault();
    handleToggle(event);
  };

  return (
    <div
      data-testid={testId}
      className={`collapse collapse-arrow bg-base-200 ${isOpen ? 'collapse-open' : ''}`}
    >
      <button
        type="button"
        className="collapse-title text-xl font-medium"
        aria-expanded={isOpen}
        aria-controls={contentId}
        onClick={handleToggle}
        onKeyDown={handleKeyDown}
        onTouchEnd={handleTouchEnd}
        tabIndex={0}
      >
        {title}
      </button>
      <div
        id={contentId}
        className="collapse-content"
        role="region"
        aria-hidden={!isOpen}
      >
        <p>{content}</p>
      </div>
    </div>
  );
};

export default CollapsibleSection;
