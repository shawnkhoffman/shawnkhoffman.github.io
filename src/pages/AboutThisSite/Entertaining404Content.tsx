import React from 'react';
import NotFound from '../404';

const Entertaining404Content: React.FC = () => {
  return (
    <div className="h-full flex flex-col p-4 sm:p-6 w-full max-w-3xl mx-auto" role="region" aria-label="404 Content">
      <div className="mb-6">
        <p className="mb-4 text-sm sm:text-base" data-testid="404-description">
          The 404 page was designed to add a touch of humor and personality to an otherwise frustrating experience.
          Instead of a generic error message, the page randomly displays amusing messages and gifs, making it more
          enjoyable for users who may have navigated to a non-existent page. This approach aims to reduce user frustration
          and create a more engaging experience.
        </p>
      </div>
      <div className="flex-grow flex flex-col items-center justify-center space-y-6">
        <div className="w-full max-w-full sm:max-w-md lg:max-w-lg bg-base-200 p-4 sm:p-6 rounded-lg shadow-md mx-auto">
          <NotFound />
        </div>
      </div>
    </div>
  );
};

export default Entertaining404Content;
