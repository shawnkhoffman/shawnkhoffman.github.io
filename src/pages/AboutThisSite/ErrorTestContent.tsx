import React, { useState, useRef, useEffect } from 'react';
import { FaChevronDown } from 'react-icons/fa';
import ErrorBoundary from '../../components/common/ErrorBoundary';

class SimulatedError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'SimulatedError';
  }
}

interface ErrorTestContentProps {
  onClose?: () => void;
}

const BuggyComponent: React.FC<{ errorType: string }> = ({ errorType }) => {
  if (errorType === 'network') {
    throw new SimulatedError('NetworkError: Failed to connect');
  } else if (errorType === 'fetch') {
    throw new SimulatedError('Failed to fetch: Error retrieving data');
  } else if (errorType === 'cors') {
    throw new SimulatedError('CORS error: Blocked by CORS policy');
  } else if (errorType === 'auth') {
    throw new SimulatedError('AuthenticationError: Invalid credentials');
  } else if (errorType === 'timeout') {
    throw new SimulatedError('TimeoutError: The request timed out');
  } else if (errorType === 'validation') {
    throw new SimulatedError('ValidationError: Form input is invalid');
  } else if (errorType === 'rateLimit') {
    throw new SimulatedError('RateLimitError: Too many requests');
  } else if (errorType === 'generic') {
    throw new SimulatedError('Boop beep boooooop!');
  }
  return <div className="text-success text-center">No Errors</div>;
};

const ErrorTestContent: React.FC<ErrorTestContentProps> = ({ onClose }) => {
  const [errorType, setErrorType] = useState<string | null>(null);
  const [selectedError, setSelectedError] = useState<string | null>(null);
  
  const resetSourceRef = useRef<'selectingError' | 'clearingError' | null>(null);
  
  const errorBoundaryRef = useRef<ErrorBoundary>(null);

  const errorTypes = [
    { label: 'Network Error', value: 'network' },
    { label: 'Fetch Error', value: 'fetch' },
    { label: 'CORS Error', value: 'cors' },
    { label: 'Authentication Error', value: 'auth' },
    { label: 'Timeout Error', value: 'timeout' },
    { label: 'Validation Error', value: 'validation' },
    { label: 'Rate Limit Error', value: 'rateLimit' },
    { label: 'Generic Error', value: 'generic' },
  ];

  const handleErrorSelection = (errorValue: string, event?: React.MouseEvent) => {
    if (event) {
      event.preventDefault();
      event.stopPropagation();
    }
    
    resetSourceRef.current = 'selectingError';
    setSelectedError(errorValue);
    setErrorType(null);
    errorBoundaryRef.current?.resetError();
    
    const detailsElement = document.querySelector('.dropdown details');
    if (detailsElement instanceof HTMLDetailsElement) {
      detailsElement.open = false;
    }
  };

  const simulateError = () => {
    if (selectedError) {
      setErrorType(selectedError);
    }
  };

  const clearError = () => {
    resetSourceRef.current = 'clearingError';
    setErrorType(null);
    errorBoundaryRef.current?.resetError();
  };

  const handleReset = () => {
    setErrorType(null);
    
    if (!resetSourceRef.current && onClose) {
      onClose();
    }

    resetSourceRef.current = null;
  };

  useEffect(() => {
    const dropdownButton = document.querySelector('[data-testid="error-dropdown-button"]');
    if (dropdownButton && dropdownButton instanceof HTMLElement) {
      dropdownButton.focus();
    }
  }, []);

  return (
    <div className="h-full flex flex-col p-4 sm:p-6 w-full max-w-3xl mx-auto">
      <div className="mb-6">
        <p className="mb-4 text-sm sm:text-base">
          This website uses an <strong>ErrorBoundary</strong> to catch and gracefully handle unexpected errors.
          When an error occurs, instead of crashing the entire site, the ErrorBoundary catches it and displays a
          relevant error message. This ensures the user experience remains intact even when problems arise.
        </p>
        <p className="mb-4 text-sm sm:text-base">
          Below you can simulate different types of errors that the ErrorBoundary can catch.
          Use the dropdown menu to select an error type, then click the "Test" button to simulate that error.
          You can clear the error afterward by using the "Clear Error" button or just select a different error type.
        </p>
      </div>

      <div className="flex-grow flex flex-col items-center justify-center space-y-6">
        <div className="dropdown dropdown-bottom mb-6 w-full max-w-xs sm:max-w-xl">
          <details className="w-full">
            <summary
              data-testid="error-dropdown-button"
              className="btn w-full flex items-center justify-between"
            >
              {errorTypes.find((error) => error.value === selectedError)?.label || 'Select Error'}
              <FaChevronDown className="ml-2" />
            </summary>
            <ul
              id="error-type-listbox"
              role="listbox"
              aria-label="Select error type"
              className="dropdown-content menu bg-base-100 rounded-box p-2 shadow mt-2 w-full z-[1]"
              data-testid="error-dropdown-list"
            >
              {errorTypes.map((errorType) => (
                <li 
                  key={errorType.value} 
                  role="option" 
                  aria-selected={selectedError === errorType.value}
                  data-testid={`error-option-${errorType.value}`}
                >
                  <a 
                    onClick={(e) => handleErrorSelection(errorType.value, e)}
                    href="#"
                    className="block w-full"
                  >
                    {errorType.label}
                  </a>
                </li>
              ))}
            </ul>
          </details>
        </div>

        <div className="flex justify-center gap-4 mb-6">
          <button 
            data-testid="test-error-button"
            className={`btn btn-sm btn-outline btn-primary ${selectedError ? '' : 'btn-disabled'}`}
            onClick={simulateError}
            disabled={!selectedError}
          >
            Test
          </button>
          <button
            data-testid="clear-error-button"
            className={`btn btn-sm btn-outline btn-accent ${errorType ? '' : 'btn-disabled'}`}
            onClick={clearError}
            disabled={!errorType}
          >
            Clear Error
          </button>
        </div>

        <div className="w-full max-w-xs sm:max-w-xl bg-base-200 p-4 sm:p-6 rounded-lg shadow-md mx-auto">
          <ErrorBoundary
            ref={errorBoundaryRef}
            showReloadButton={false}
            isTestError={true}
            onReset={handleReset}
          >
            {errorType ? (
              <div className="text-error text-center">
                <BuggyComponent errorType={errorType} />
              </div>
            ) : (
              <div className="text-success text-center">Please select an error to test.</div>
            )}
          </ErrorBoundary>
        </div>
      </div>
      <div className="pb-10"></div>
    </div>
  );
};

export default ErrorTestContent;
