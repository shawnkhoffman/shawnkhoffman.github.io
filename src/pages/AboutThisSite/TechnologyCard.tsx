import React from 'react';

interface TechnologyCardProps {
  icon: React.ReactNode;
  title: string;
  description: string;
  link: string;
}

const TechnologyCard: React.FC<TechnologyCardProps> = ({ icon, title, description, link }) => {
  return (
    <a
      href={link}
      target="_blank"
      rel="noopener noreferrer"
      className="flex flex-col items-center text-center hover:bg-base-200 p-4 rounded-lg transition duration-300"
      aria-label={`Learn more about ${title}`}
    >
      <span aria-hidden="true">{icon}</span>
      <h3 className="font-semibold">{title}</h3>
      <p className="text-sm">{description}</p>
    </a>
  );
};

export default TechnologyCard;