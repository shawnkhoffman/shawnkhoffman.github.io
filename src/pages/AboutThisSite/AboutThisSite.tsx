import React, { useState } from 'react';
import { FaReact, FaGithub, FaCode, FaLayerGroup, FaTools, FaBug, FaWrench, FaLaughBeam } from 'react-icons/fa';
import TechnologyCard from './TechnologyCard';
import Modal from '../../components/common/Modal';
import ErrorTestContent from './ErrorTestContent';
import Entertaining404Content from './Entertaining404Content';

interface Feature {
  id: string;
  category: string;
  icon: React.ReactElement;
  content: React.ReactElement;
  title?: string;
}

const AboutThisSite: React.FC = () => {
  const [currentFeatureIndex, setCurrentFeatureIndex] = useState<number | null>(null);
  const [isFeatureModalOpen, setIsFeatureModalOpen] = useState(false);
  const [isFeatureModalExpanded, setIsFeatureModalExpanded] = useState(true);
  const [triggerOverflowCheck, setTriggerOverflowCheck] = useState(0);

  const modalFeatures = [
    {
      id: 'error-handling',
      category: 'Error Handling',
      icon: <FaBug className="text-4xl text-error mb-4" aria-hidden="true" />,
      content: (
        <ErrorTestContent onClose={() => setIsFeatureModalOpen(false)} />
      ),
    },
    {
      id: 'entertaining-404',
      category: 'Entertaining 404 Page',
      icon: <FaLaughBeam className="text-4xl text-warning mb-4" aria-hidden="true" />,
      content: <Entertaining404Content />
    },
  ];

  const allFeatures = [
    ...modalFeatures,
    {
      id: 'coming-soon',
      category: 'Coming Soon',
      icon: <FaWrench className="text-4xl text-base-content mb-4" aria-hidden="true" />,
      content: <p>This feature is currently under development.</p>,
    },
  ];

  const handleFeatureClick = (feature: Feature) => {
    if (feature.id === 'coming-soon') return;
    setCurrentFeatureIndex(modalFeatures.findIndex(f => f.id === feature.id));
    setIsFeatureModalOpen(true);
  };

  const closeFeatureModal = () => {
    setIsFeatureModalOpen(false);
  };

  const handleFeatureNext = () => {
    setCurrentFeatureIndex((prevIndex) =>
      prevIndex === null
        ? 0
        : (prevIndex + 1) % modalFeatures.length
    );
  };

  const handleFeaturePrevious = () => {
    setCurrentFeatureIndex((prevIndex) =>
      prevIndex === null
        ? modalFeatures.length - 1
        : (prevIndex - 1 + modalFeatures.length) % modalFeatures.length
    );
  };

  const toggleFeatureModalExpand = () => {
    setIsFeatureModalExpanded(!isFeatureModalExpanded);
    setTriggerOverflowCheck(prev => prev + 1);
  };

  return (
    <main aria-label="About This Site Section" className="flex flex-col items-center justify-center w-full p-6">
      {/* Heading aligned with About page styling */}
      <section 
        aria-labelledby="about-this-site-heading"
        className="w-full max-w-3xl text-center mb-10 pt-4"
      >
        <h1 
          id="about-this-site-heading" 
          className="text-4xl font-bold mb-4"
        >
          About This Site
        </h1>
      </section>

      {/* Technologies Section */}
      <section className="w-full max-w-5xl mb-10 p-6 bg-base-100 rounded-lg shadow-lg">
        <h2 className="text-2xl font-semibold mb-6 text-center">Technologies Used</h2>
        <div className="grid grid-cols-2 sm:grid-cols-3 gap-6">
          <TechnologyCard
            icon={<FaReact className="text-4xl text-info mb-2" aria-hidden="true" />}
            title="React"
            description="Fast, component-based UI development"
            link="https://reactjs.org"
          />
          <TechnologyCard
            icon={<FaCode className="text-4xl text-info mb-2" aria-hidden="true" />}
            title="TypeScript"
            description="Type-safe development environment"
            link="https://www.typescriptlang.org"
          />
          <TechnologyCard
            icon={<FaLayerGroup className="text-4xl text-info mb-2" aria-hidden="true" />}
            title="Tailwind CSS & DaisyUI"
            description="Utility-first CSS framework"
            link="https://daisyui.com"
          />
          <TechnologyCard
            icon={<FaGithub className="text-4xl text-info mb-2" aria-hidden="true" />}
            title="GitHub Pages"
            description="For static site deployment"
            link="https://pages.github.com"
          />
          <TechnologyCard
            icon={<FaTools className="text-4xl text-info mb-2" aria-hidden="true" />}
            title="Vite"
            description="Fast development builds"
            link="https://vitejs.dev"
          />
        </div>
      </section>

      {/* Site Features Section */}
      <section className="w-full max-w-5xl mb-10 p-6 bg-base-100 rounded-lg shadow-lg">
        <h2 className="text-2xl font-semibold mb-6 text-center">Site Features</h2>
        <div className="grid grid-cols-1 md:grid-cols-2 lg:grid-cols-3 gap-8">
          {allFeatures.map((feature) => (
            <div key={feature.id} className="flex flex-col items-center">
              {feature.icon}
              <h3 className="font-semibold mb-2">{feature.category}</h3>
              <button
                className="btn btn-sm mt-4"
                onClick={() => handleFeatureClick(feature)}
                aria-label={`Learn more about ${feature.category}`}
                data-testid={`learn-more-${feature.id}`}
                disabled={feature.id === 'coming-soon'}
              >
                {feature.id === 'coming-soon' ? 'Under Construction' : 'Learn More'}
              </button>
            </div>
          ))}
        </div>
      </section>

      {currentFeatureIndex !== null && (
        <Modal
          isOpen={isFeatureModalOpen}
          onClose={closeFeatureModal}
          title={modalFeatures[currentFeatureIndex].category}
          content={modalFeatures[currentFeatureIndex].content}
          onNext={handleFeatureNext}
          onPrevious={handleFeaturePrevious}
          totalPages={modalFeatures.length}
          currentPage={currentFeatureIndex}
          isExpanded={isFeatureModalExpanded}
          onToggleExpand={toggleFeatureModalExpand}
          triggerOverflowCheck={triggerOverflowCheck}
        />
      )}

      {/* GitHub Source Code Section */}
      <section className="flex justify-center items-center mt-6 mb-12">
        <button
          className="btn btn-primary inline-flex items-center"
          onClick={() => window.open('https://github.com/shawnkhoffman/shawnkhoffman.github.io', '_blank')}
          aria-label="View the source code on GitHub"
        >
          <FaGithub className="mr-2" aria-hidden="true" />
          View Source Code
        </button>
      </section>
    </main>
  );
};

export default AboutThisSite;
