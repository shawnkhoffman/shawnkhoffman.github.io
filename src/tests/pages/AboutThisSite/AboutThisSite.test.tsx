import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import { describe, it, expect, vi } from 'vitest';
import AboutThisSite from '../../../pages/AboutThisSite/AboutThisSite';

vi.mock('../../../components/common/Modal', () => ({
  default: ({ isOpen, content }: { isOpen: boolean; content: React.ReactNode }) => 
    isOpen ? <div data-testid="modal">{content}</div> : null
}));

vi.mock('../../../pages/AboutThisSite/TechnologyCard', () => ({
  default: ({ title }: { title: string }) => 
    <div data-testid="technology-card" role="link" aria-label={`Learn more about ${title}`}>{title}</div>
}));

describe('AboutThisSite', () => {
  describe('Rendering', () => {
    it('renders the page', () => {
      const { container } = render(<AboutThisSite />);
      expect(container).toBeInTheDocument();
      expect(container).toMatchSnapshot();
    });
  });

  describe('Structure', () => {
    it('contains main content sections', () => {
      render(<AboutThisSite />);
      const mainHeading = screen.getAllByRole('heading', { level: 1 })[0];
      const subHeadings = screen.getAllByRole('heading', { level: 2 });
      
      expect(mainHeading).toBeInTheDocument();
      expect(subHeadings.length).toBeGreaterThan(0);
    });

    it('displays interactive elements', () => {
      render(<AboutThisSite />);
      expect(screen.getAllByRole('button')).not.toHaveLength(0);
      expect(screen.getAllByTestId('technology-card')).not.toHaveLength(0);
    });
  });

  describe('Interactions', () => {
    it('handles modal interactions', () => {
      render(<AboutThisSite />);
      
      const buttons = screen.getAllByTestId('learn-more-error-handling');
      if (buttons.length === 0) {
        throw new Error('No error handling buttons found');
      }
      
      const button = buttons[0] as HTMLButtonElement;
      if (!button) {
        throw new Error('Error handling button not found');
      }
      
      fireEvent.click(button);
      const modal = screen.getByTestId('modal');
      expect(modal).toBeInTheDocument();
    });
  });

  describe('Accessibility', () => {
    it('has proper accessibility attributes', () => {
      render(<AboutThisSite />);
      const buttons = screen.getAllByRole('button');
      const links = screen.getAllByRole('link');
      
      buttons.forEach(button => {
        const hasAriaLabel = button.hasAttribute('aria-label');
        const hasTextContent = button.textContent ? button.textContent.trim().length > 0 : false;
        expect(hasAriaLabel || hasTextContent).toBe(true);
      });

      links.forEach(link => {
        expect(link).toHaveAttribute('aria-label');
      });
    });
  });
});
