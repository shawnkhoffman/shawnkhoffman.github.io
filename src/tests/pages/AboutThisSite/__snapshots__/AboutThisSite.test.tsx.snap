// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AboutThisSite > Accessibility > has proper accessibility attributes 1`] = `
<div>
  <main
    aria-label="About This Site Section"
    class="flex flex-col items-center justify-center w-full p-6"
  >
    <section
      aria-labelledby="about-this-site-heading"
      class="w-full max-w-3xl text-center mb-10 pt-4"
    >
      <h1
        class="text-4xl font-bold mb-4"
        id="about-this-site-heading"
      >
        About This Site
      </h1>
    </section>
    <section
      class="w-full max-w-5xl mb-10 p-6 bg-base-100 rounded-lg shadow-lg"
    >
      <h2
        class="text-2xl font-semibold mb-6 text-center"
      >
        Technologies Used
      </h2>
      <div
        class="grid grid-cols-2 sm:grid-cols-3 gap-6"
      >
        <div
          aria-label="Learn more about React"
          data-testid="technology-card"
          role="link"
        >
          React
        </div>
        <div
          aria-label="Learn more about TypeScript"
          data-testid="technology-card"
          role="link"
        >
          TypeScript
        </div>
        <div
          aria-label="Learn more about Tailwind CSS & DaisyUI"
          data-testid="technology-card"
          role="link"
        >
          Tailwind CSS & DaisyUI
        </div>
        <div
          aria-label="Learn more about GitHub Pages"
          data-testid="technology-card"
          role="link"
        >
          GitHub Pages
        </div>
        <div
          aria-label="Learn more about Vite"
          data-testid="technology-card"
          role="link"
        >
          Vite
        </div>
      </div>
    </section>
    <section
      class="w-full max-w-5xl mb-10 p-6 bg-base-100 rounded-lg shadow-lg"
    >
      <h2
        class="text-2xl font-semibold mb-6 text-center"
      >
        Site Features
      </h2>
      <div
        class="grid grid-cols-1 md:grid-cols-2 lg:grid-cols-3 gap-8"
      >
        <div
          class="flex flex-col items-center"
        >
          <svg
            aria-hidden="true"
            class="text-4xl text-error mb-4"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 512 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M511.988 288.9c-.478 17.43-15.217 31.1-32.653 31.1H424v16c0 21.864-4.882 42.584-13.6 61.145l60.228 60.228c12.496 12.497 12.496 32.758 0 45.255-12.498 12.497-32.759 12.496-45.256 0l-54.736-54.736C345.886 467.965 314.351 480 280 480V236c0-6.627-5.373-12-12-12h-24c-6.627 0-12 5.373-12 12v244c-34.351 0-65.886-12.035-90.636-32.108l-54.736 54.736c-12.498 12.497-32.759 12.496-45.256 0-12.496-12.497-12.496-32.758 0-45.255l60.228-60.228C92.882 378.584 88 357.864 88 336v-16H32.666C15.23 320 .491 306.33.013 288.9-.484 270.816 14.028 256 32 256h56v-58.745l-46.628-46.628c-12.496-12.497-12.496-32.758 0-45.255 12.498-12.497 32.758-12.497 45.256 0L141.255 160h229.489l54.627-54.627c12.498-12.497 32.758-12.497 45.256 0 12.496 12.497 12.496 32.758 0 45.255L424 197.255V256h56c17.972 0 32.484 14.816 31.988 32.9zM257 0c-61.856 0-112 50.144-112 112h224C369 50.144 318.856 0 257 0z"
            />
          </svg>
          <h3
            class="font-semibold mb-2"
          >
            Error Handling
          </h3>
          <button
            aria-label="Learn more about Error Handling"
            class="btn btn-sm mt-4"
            data-testid="learn-more-error-handling"
          >
            Learn More
          </button>
        </div>
        <div
          class="flex flex-col items-center"
        >
          <svg
            aria-hidden="true"
            class="text-4xl text-warning mb-4"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 496 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M248 8C111 8 0 119 0 256s111 248 248 248 248-111 248-248S385 8 248 8zm24 199.4c3.3-42.1 32.2-71.4 56-71.4s52.7 29.3 56 71.4c.7 8.6-10.8 11.9-14.9 4.5l-9.5-17c-7.7-13.7-19.2-21.6-31.5-21.6s-23.8 7.9-31.5 21.6l-9.5 17c-4.2 7.4-15.8 4.1-15.1-4.5zm-160 0c3.3-42.1 32.2-71.4 56-71.4s52.7 29.3 56 71.4c.7 8.6-10.8 11.9-14.9 4.5l-9.5-17c-7.7-13.7-19.2-21.6-31.5-21.6s-23.8 7.9-31.5 21.6l-9.5 17c-4.3 7.4-15.8 4-15.1-4.5zM398.9 306C390 377 329.4 432 256 432h-16c-73.4 0-134-55-142.9-126-1.2-9.5 6.3-18 15.9-18h270c9.6 0 17.1 8.4 15.9 18z"
            />
          </svg>
          <h3
            class="font-semibold mb-2"
          >
            Entertaining 404 Page
          </h3>
          <button
            aria-label="Learn more about Entertaining 404 Page"
            class="btn btn-sm mt-4"
            data-testid="learn-more-entertaining-404"
          >
            Learn More
          </button>
        </div>
        <div
          class="flex flex-col items-center"
        >
          <svg
            aria-hidden="true"
            class="text-4xl text-content mb-4"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 512 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M507.73 109.1c-2.24-9.03-13.54-12.09-20.12-5.51l-74.36 74.36-67.88-11.31-11.31-67.88 74.36-74.36c6.62-6.62 3.43-17.9-5.66-20.16-47.38-11.74-99.55.91-136.58 37.93-39.64 39.64-50.55 97.1-34.05 147.2L18.74 402.76c-24.99 24.99-24.99 65.51 0 90.5 24.99 24.99 65.51 24.99 90.5 0l213.21-213.21c50.12 16.71 107.47 5.68 147.37-34.22 37.07-37.07 49.7-89.32 37.91-136.73zM64 472c-13.25 0-24-10.75-24-24 0-13.26 10.75-24 24-24s24 10.74 24 24c0 13.25-10.75 24-24 24z"
            />
          </svg>
          <h3
            class="font-semibold mb-2"
          >
            Coming Soon
          </h3>
          <button
            aria-label="Learn more about Coming Soon"
            class="btn btn-sm mt-4"
            data-testid="learn-more-coming-soon"
            disabled=""
          >
            Under Construction
          </button>
        </div>
      </div>
    </section>
    <section
      class="flex justify-center items-center mt-6 mb-12"
    >
      <button
        aria-label="View the source code on GitHub"
        class="btn btn-primary inline-flex items-center"
      >
        <svg
          aria-hidden="true"
          class="mr-2"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 496 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"
          />
        </svg>
        View Source Code
      </button>
    </section>
  </main>
</div>
`;
